File: src/skaters/conformal.py
"""Online conformal recalibration of a distributional skater.

A skater's predictive distribution can be the right *location* and *scale*
yet the wrong *shape* — a Gaussian leaf under-covers the tails and
over-covers the centre of heavy-tailed residuals, and Bayesian model
averaging washes kurtosis out. Conformal recalibration fixes this from the
outside: it replaces the predictive shape with the empirical distribution
of recent standardized residuals.

This is split conformal prediction on a rolling window, run online:

    z_t = (y_t - mu_t) / sigma_t            standardized residual
    F_calibrated(.) = mu + sigma * Z_emp    Z_emp = recent {z}

The base's mu and sigma still carry the mean forecast and the (possibly
heteroskedastic) scale; conformal supplies the shape non-parametrically, so
heavy tails, skew, and miscalibration are corrected and the predictive
intervals attain approximately their nominal coverage. Pure Python; the
result is still a Dist (a small Gaussian mixture at the empirical
quantiles), so it composes with everything else.
"""

from __future__ import annotations
import math
from collections import deque
from skaters.dist import Dist


def conformal(base, k: int = 1, window: int = 250, min_obs: int = 30,
              n_points: int = 40, bandwidth: float = 1.0):
    """Wrap a skater so its intervals are conformally calibrated online.

    Args:
        base: any skater callable (y, state) -> (list[Dist], state).
        k: forecast horizon (must match base).
        window: rolling calibration window of standardized residuals.
            Smaller = adapts faster to non-stationarity; larger = smoother.
        min_obs: pass the base prediction through unchanged until this many
            residuals have accumulated.
        n_points: number of empirical-quantile components in the calibrated
            Dist (a KDE-style Gaussian mixture).

    Returns:
        A skater callable producing calibrated list[Dist].
    """

    def _skater(y: float, state: dict | None) -> tuple[list[Dist], dict]:
        if state is None:
            state = {"base": None, "z": deque(maxlen=window), "pending": None}

        # Resolve the previous one-step prediction into a standardized residual.
        if state["pending"] is not None:
            mu_prev, sigma_prev = state["pending"]
            if sigma_prev > 0:
                state["z"].append((y - mu_prev) / sigma_prev)

        dists, state["base"] = base(y, state["base"])

        d1 = dists[0]
        state["pending"] = (d1.mean, d1.std)

        z = list(state["z"])
        if len(z) < min_obs:
            return dists, state  # warm up on the base distribution

        # Bandwidth (Silverman) for smoothing the empirical residual law.
        n = len(z)
        zmean = sum(z) / n
        zvar = sum((v - zmean) ** 2 for v in z) / n
        zstd = math.sqrt(zvar) if zvar > 0 else 1.0
        h = bandwidth * 0.9 * zstd * n ** (-0.2)

        zs = sorted(z)
        out = []
        for d in dists:
            mu = d.mean
            sigma = d.std
            if sigma <= 0:
                out.append(d)
                continue
            comps = []
            for j in range(n_points):
                q = (j + 0.5) / n_points
                zq = zs[min(int(q * n), n - 1)]
                comps.append((1.0 / n_points, mu + sigma * zq, sigma * h))
            out.append(Dist(comps))
        return out, state

    _skater.__name__ = f"conformal({getattr(base, '__name__', '?')})"
    return _skater

File: docs/js/skaters/conformal.mjs
// Online conformal recalibration — JS port of skaters/conformal.py.
//
// Replaces a skater's predictive *shape* with the empirical distribution of
// recent standardized residuals (split conformal on a rolling window),
// keeping the base's location and scale. Returns a calibrated Dist.

import { Dist } from "./dist.mjs";

export function conformal(base, k = 1, window = 250, minObs = 30, nPoints = 40, bandwidth = 1.0) {
  function _skater(y, state) {
    if (state === null || state === undefined) {
      state = { base: null, z: [], pending: null };
    }

    if (state.pending !== null) {
      const [muPrev, sigmaPrev] = state.pending;
      if (sigmaPrev > 0) {
        state.z.push((y - muPrev) / sigmaPrev);
        if (state.z.length > window) state.z.shift();
      }
    }

    const [dists, baseState] = base(y, state.base);
    state.base = baseState;

    const d1 = dists[0];
    state.pending = [d1.mean, d1.std];

    const z = state.z;
    if (z.length < minObs) return [dists, state];

    const n = z.length;
    let zmean = 0.0;
    for (const v of z) zmean += v;
    zmean /= n;
    let zvar = 0.0;
    for (const v of z) zvar += (v - zmean) * (v - zmean);
    zvar /= n;
    const zstd = zvar > 0 ? Math.sqrt(zvar) : 1.0;
    const h = bandwidth * 0.9 * zstd * Math.pow(n, -0.2);

    const zs = z.slice().sort((a, b) => a - b);
    const out = [];
    for (const d of dists) {
      const mu = d.mean;
      const sigma = d.std;
      if (sigma <= 0) {
        out.push(d);
        continue;
      }
      const comps = [];
      for (let j = 0; j < nPoints; j++) {
        const q = (j + 0.5) / nPoints;
        const zq = zs[Math.min(Math.floor(q * n), n - 1)];
        comps.push([1.0 / nPoints, mu + sigma * zq, sigma * h]);
      }
      out.push(new Dist(comps));
    }
    return [out, state];
  }
  _skater.skaterName = `conformal(${base.skaterName || "?"})`;
  return _skater;
}
